arch/arc/core/arc_smp.c

```diff
@@ -98,32 +98,55 @@ static void sched_ipi_handler(void *unused)
 /**
  * @brief Check whether need to do thread switch in isr context
  *
- * @details u64_t is used to let compiler use (r0, r1) as return register.
- *  use register r0 and register r1 as return value, r0 has
- *  new thread, r1 has old thread. If r0 == 0, it means no thread switch.
+ * @param old, current thread pointer will be saved in old
+ *
+ * @return new thread pointer
+ *
+ * @details in isr, two values are required, the new thread pointer and
+ * old thread pointer, if new thread pointer is not null, a thread switch
+ * is required, old thread pointer will be used to save context; if new
+ * thread pointer is null, no thread switch
  */
-u64_t z_arc_smp_switch_in_isr(void)
+struct k_thread *z_arc_smp_switch_in_isr(struct k_thread **old)
 {
-	u64_t ret = 0;
-	u32_t new_thread;
-	u32_t old_thread;
+	struct k_thread *new_thread, *old_thread;
+
+	old_thread = _current;
+
+	/* Null out the switch handle, see wait_for_switch() above.
```

Comment:
Looks like this comment is copy & pasted from kswap.h or so, there is no "wait_for_switch() above".
Do we need this all this "switch_handle-ing" in the architecture specific code? No way to use the generic code, or call generic (inline) functions for this? (I'm a bit concerned with maintainability).

Reply:
yes, it's copy from kswap.h. For ARC, thread switch in interrupt context cannot reuse the arch_switch， so we have to add handling for interrupt case. For ARM,  pending_sv exception is the only entry of thread switch. This make things easier